crates/noirc_frontend/src/node_interner.rs

```diff
@@ -71,6 +73,11 @@ pub struct NodeInterner {
     // We'd just lookup their methods as needed through the NodeInterner.
     traits: HashMap<TraitId, Shared<Trait>>,
 
+    // Trait implementation map
+    // For each type that implements a given Trait ( corresponding TraitId), there should be an entry here
+    // The purpose for this hashmap is to detect duplication of trait implementations ( if any )
+    trait_implementaions: HashMap<(Type, TraitId), Ident>,
```

Comment:
I think we'll need to refine this further in the future to catch duplicate generic implementations. E.g. `impl<U> Foo for Bar<U>` and `impl<U> Foo for Bar<U>` are equivalent but would be different keys in this map.

Reply:
Yeah, in my local branch I'm currently implementing generic traits, the signature looks like:
```rust
struct TraitImpl {
    pub methods: Vec<FuncId>, // list of the concrete method implementation
    ...
}

struct TraitImplKey {
    typ: Type,
    trait_id: TraitId,
    generics: Vec<Type<
}

trait_implementaions: HashMap<(Type, TraitId), Shared<TraitImpl>>,